src/api/java/IDecRefQueue.java

```diff
@@ -34,38 +42,31 @@ protected IDecRefQueue(int move_limit)
     {
     	m_move_limit = move_limit;
     }
- 
-    public void setLimit(int l) { m_move_limit = l; }
- 
-    protected abstract void incRef(Context ctx, long obj);
 
+    /**
+     * An implementation of this method should decrement the reference on a
+     * given native object.
+     * This function should be always called on the {@code ctx} thread.
+     *
+     * @param ctx Z3 context.
+     * @param obj Pointer to a Z3 object.
+     */
     protected abstract void decRef(Context ctx, long obj);
 
-    protected void incAndClear(Context ctx, long o)
-    {
-        incRef(ctx, o);
-        if (m_queue.size() >= m_move_limit)
-            clear(ctx);
-    }
-
-    protected void add(long o)
-    {
-        if (o == 0)
-            return;
+    public void storeReference(Context ctx, T obj) {
+        PhantomReference<T> ref = new PhantomReference<>(obj, referenceQueue);
+        referenceMap.put(ref, obj.getNativeObject());
 
-        synchronized (m_lock)
-        {
-            m_queue.add(o);
-        }
+        // TODO: use move_limit, somehow get the size of referenceQueue.
+        clear(ctx);
     }
 
     protected void clear(Context ctx)
```

Comment:
Sorry, I had some problems code-reviewing based on the diffs only. It got clearer from going to your fork and looking at full files: you call "clear" whenever allocating a new object and closing the context. So I assume the weak pointers become available in the queue using "poll" whenever the GC kicks in on their owner objects. I also assume that poll and remove are synchronized and therefore don't have thread issues with the GC. Indeed this solution would enable the dual purpose collection of Z3Objects.